Core/LocalizationManager.cs

```diff
@@ -156,16 +156,63 @@ public IEnumerable<Object> GetScriptResources(CultureInfo culture)
 
 		public IEnumerable<ILocalizedObject> GetAll(CultureInfo culture)
 		{
-			var lst = LocalizationCache.Get<IEnumerable<ILocalizedObject>>(culture.Name);
+			IEnumerable<ILocalizedObject> lst = null;
+			if (_initialBundlesCount > 0)
+			{
+				List<ILocalizedObject>[] bundles = new List<ILocalizedObject>[_initialBundlesCount];
+				for (int i = 0; i < _initialBundlesCount; i++)
+				{
+					bundles[i] = LocalizationCache.Get<List<ILocalizedObject>>(GetBundleName(culture, i));
+					if (bundles[i] == null)
+						break;
+				}
+				if (!bundles.Any(l => l == null))
+				{
+					var lstCombined = new List<ILocalizedObject>();
+					foreach (var bundle in bundles)
+						lstCombined.AddRange(bundle);
+					lst = lstCombined;
+				}
+			}
 			if (lst == null || lst.Count() == 0)
 			{
 				lst = Repository.GetAll(culture).ToList();
-				LocalizationCache.Insert(culture.Name, lst);
+				if (_initialBundlesCount <= 0)
+					_initialBundlesCount = lst.Count() * 1024 / (1024 * 1024) * 4;
+				List<ILocalizedObject>[] bundles = new List<ILocalizedObject>[_initialBundlesCount];
+				for (int i = 0; i < _initialBundlesCount; i++)
+					bundles[i] = new List<ILocalizedObject>();
+				foreach (var obj in lst)
+					bundles[GetBundleIndex(obj.Hash)].Add(obj);
+
+				for (int i = 0; i < _initialBundlesCount; i++)
+					LocalizationCache.Insert(GetBundleName(culture, i), bundles[i]);
 			}
 
 			return lst;
 		}
 
+		private static string GetBundleName(CultureInfo culture, int bundleIndex)
+		{
+			return string.Format("{0}_bundle{1}", culture.Name, bundleIndex);
+		}
+
+		private int GetBundleIndex(string Hash)
+		{
+			int index = Hash.GetHashCode();
+			if (index < 0)
+				index = -index;
+			index %= _initialBundlesCount;
+			return index;
+		}
+
+		public IEnumerable<ILocalizedObject> GetCachedListForHash(CultureInfo culture, string Hash)

```

Comment:
This method is used to get LocalizedObject by hash, so we doesn't have object itself to call it.